x/authenticator/authenticator/any_of.go

```diff
@@ -74,70 +68,41 @@ func (aoa AnyOfAuthenticator) Initialize(data []byte) (iface.Authenticator, erro
 }
 
 func (aoa AnyOfAuthenticator) Authenticate(ctx sdk.Context, request iface.AuthenticationRequest) error {
-	baseId := request.AuthenticatorId
-
-	var err error
-	for id, auth := range aoa.SubAuthenticators {
-		request.AuthenticatorId = baseId + "." + strconv.Itoa(id)
-		err = auth.Authenticate(ctx, request)
-		ctx.Logger().Debug("AnyOfAuthenticator.Authenticate", "request.AuthenticatorId", request.AuthenticatorId, "err", err)
-		// early return ok if any of the sub-authenticators return ok
-		if err == nil {
-			return nil
-		}
+	if len(aoa.SubAuthenticators) == 0 {
+		return errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "no sub-authenticators provided")
 	}
-	return errorsmod.Wrap(sdkerrors.ErrUnauthorized, "all sub-authenticators failed to authenticate")
-}
 
-func (aoa AnyOfAuthenticator) Track(ctx sdk.Context, account sdk.AccAddress, msg sdk.Msg, msgIndex uint64,
-	authenticatorId string) error {
-	for id, auth := range aoa.SubAuthenticators {
-		err := auth.Track(ctx, account, msg, msgIndex, authenticatorId+"."+strconv.Itoa(id))
-		if err != nil {
-			return err
-		}
+	err := subHandleRequest(
+		ctx, request, aoa.SubAuthenticators, requireAnyPass,
+		func(auth iface.Authenticator, ctx sdk.Context, request iface.AuthenticationRequest) error {
+			return auth.Authenticate(ctx, request)
+		},
+	)
+
+	if err != nil {
+		return errorsmod.Wrap(sdkerrors.ErrUnauthorized, "all sub-authenticators failed to authenticate")
 	}
+
 	return nil
 }
 
+func (aoa AnyOfAuthenticator) Track(ctx sdk.Context, account sdk.AccAddress, msg sdk.Msg, msgIndex uint64, authenticatorId string) error {
+	return subTrack(ctx, account, msg, msgIndex, authenticatorId, aoa.SubAuthenticators)
+}
+
 func (aoa AnyOfAuthenticator) ConfirmExecution(ctx sdk.Context, request iface.AuthenticationRequest) error {
-	baseId := request.AuthenticatorId
-	for id, auth := range aoa.SubAuthenticators {
-		request.AuthenticatorId = baseId + "." + strconv.Itoa(id)
-		err := auth.ConfirmExecution(ctx, request)
-		if err != nil {
-			return err
-		}
-	}
-	return nil
+	return subHandleRequest(
+		ctx, request, aoa.SubAuthenticators, requireAllPass,
```

Comment:
```suggestion
                // When doing confirm on an AnyOf, we still want to requireAllPass. This is to limit the risk
                // when someone adds authenticators that only do checks on confirm in an anyOf
		ctx, request, aoa.SubAuthenticators, requireAllPass,
```

Reply:
Added a comment with the explanation... but maybe this is not something we want and it's the user's responsibility to add the correct authenticators